In [26]:
import pandas as pd
import matplotlib.pyplot as plt

In [27]:
shooters = pd.read_csv('Data/Stanford_MSA_Database.csv')

OK, so there's lots of things we can do with the dataset given that it's a full 335 rows with 55 columns of descriptive text data each. Let's start by coming up with some basic questions we want answered:
- What kinds of weapons are most popular, and where?
- Where are guns most readily accessible in the country?
- How do warning signs change with age and other factors?
- Does lethality change with age?
- Have shootings become, on the whole, more lethal over time?
- How do the number of fatalities compare to the total number of victims?
- Has the number of guns used in a shooting changed on average over time?
- What happens to the shooter, and how does that change with age and race? Does their fate change substantially in incidences of school shootings?
- How do the motivations of mass shooters differ with geography?
- Does military experience alter the motive?

In [28]:
shooters

CaseID  \
0         1   
1         2   
2         3   
3         4   
4         5   
5         6   
6         7   
7         8   
8         9   
9        10   
10       11   
11       12   
12       13   
13       14   
14       15   
15       16   
16       17   
17       18   
18       19   
19       20   
20       21   
21       22   
22       23   
23       24   
24       25   
25       26   
26       27   
27       28   
28       29   
29       30   
..      ...   
305     315   
306     316   
307     317   
308     318   
309     319   
310     320   
311     321   
312     322   
313     323   
314     324   
315     325   
316     326   
317     327   
318     328   
319     329   
320     330   
321     331   
322     332   
323     333   
324     334   
325     335   
326     336   
327     337   
328     338   
329     339   
330     340   
331     341   
332     342   
333     343   
334     344   

                                                                     Title  \
0                                            University of Texas at Austin   
1                                               Rose-Mar College of Beauty   
2                                             New Orleans Police Shootings   
3                                           Clara Barton Elementary School   
4                                                        Olean High School   
5                                     Los Angeles Computer Learning Center   
6                                                      Cal State Fullerton   
7                                       Grover Cleveland Elementary School   
8                                             University of South Carolina   
9                                                       Valley High School   
10                                                   Welding shop in Miami   
11                                                 Wah Mee Club in Seattle   
12                                                    Johnston Post Office   
13                                           49th Street Elementary School   
14                                                  Ianni's Club in Dallas   
15                                     McDonald's restaurant in San Ysidro   
16                                              Goddard Junior High School   
17                                                     Atlanta Post Office   
18                                              New York Technical College   
19                                         Post office in Edmond, Oklahoma   
20                                               Fergus County High School   
21                                               Palm Bay Shopping Centers   
22                                               Pinellas Park High School   
23                                Electromagnetic Systems Laboratory (ESL)   
24                                         Hubbard Woods Elementary School   
25                                  Comet Auto Parts and Montefiore School   
26                                               Oakland Elementary School   
27                                        New Orleans Downtown Post Office   
28                                             Cleveland Elementary School   
29                                                Orange Glenn Post Office   
..                                                                     ...   
305                                         Forestville, Maryland Drive-by   
306                                                  Ferguson, MO Drive by   
307                                              Appling, GA family murder   
308                                               Auburn, AL club shootout   
309                                  Nashville Music City Central shooting   
310                                        Murfreesboro, TN party shooting   
311                                  Florida Ex-wife, ex-girlfriend murder   
312                             

In [75]:
len(shooters)

335

In [29]:
list(shooters)

['CaseID',
 'Title',
 'Location',
 'City',
 'State',
 'Latitude',
 'Longitude',
 'Number of Civilian Fatalities',
 'Number of Civilian Injured',
 'Number of Enforcement Fatalities',
 'Number of Enforcement Injured',
 'Total Number of Fatalities',
 'Total Number of Victims',
 'Description',
 'Date',
 'Day of Week',
 'Date - Detailed',
 'Shooter Name',
 'Number of shooters',
 'Shooter Age(s)',
 'Average Shooter Age',
 'Shooter Sex',
 'Shooter Race',
 'Type of Gun - Detailed',
 'Type of Gun - General',
 'Number of Shotguns',
 'Number of Rifles',
 'Number of Handguns',
 'Total Number of Guns',
 'Number of Automatic Guns',
 'Number of Semi-Automatic Guns',
 'Fate of Shooter at the scene',
 'Fate of Shooter',
 "Shooter's Cause of Death",
 'School Related',
 'Place Type',
 'Relationship to Incident Location',
 'Targeted Victim/s - Detailed',
 'Targeted Victim/s - General',
 'Possible Motive - Detailed',
 'Possible Motive - General',
 'History of Mental Illness - Detailed',
 'History of Mental

I'm curious what's going to happen if I run a market basket analysis. Let's clean up the data a little more and see what happens. I'm working off of the following tutorial:
http://pbpython.com/market-basket-analysis.html

In [30]:
cleaned_shooters = shooters.copy(deep=True)

#Clean gun types
cleaned_shooters['Type of Gun - General'] = cleaned_shooters['Type of Gun - General'].str.strip()
cleaned_shooters['Type of Gun - General'] = cleaned_shooters['Type of Gun - General'].str.lower()
cleaned_shooters['Type of Gun - General'].unique()

array(['multiple guns', 'handgun', 'shotgun', 'rifle', 'unknown',
       'semi-automatic rifle', '9-mm'], dtype=object)

In [31]:
#Clean gun counts
guncount_types = ['Number of Shotguns','Number of Handguns','Number of Rifles']
for guncount_type in guncount_types:
    cleaned_shooters[guncount_type].replace('Unknown', 0, inplace=True)

In [32]:
#Clean military experience
cleaned_shooters['Military Experience'] = cleaned_shooters['Military Experience'].str.lower()
cleaned_shooters['Military Experience'].replace('yes', True, inplace=True)
cleaned_shooters['Military Experience'].replace(['no', 'unknown'], False, inplace=True)

In [33]:
#Set civilian + law enforcement casualties, injuries to bools
cleaned_shooters['Number of Civilian Fatalities'] = cleaned_shooters['Number of Civilian Fatalities'] > 0
cleaned_shooters['Number of Civilian Injured'] = cleaned_shooters['Number of Civilian Injured'] > 0
cleaned_shooters['Number of Enforcement Fatalities'] = cleaned_shooters['Number of Enforcement Fatalities'] > 0
cleaned_shooters['Number of Enforcement Injured'] = cleaned_shooters['Number of Enforcement Injured'] > 0
cleaned_shooters['High Civilian Casualties'] = (shooters['Number of Civilian Fatalities'] + shooters['Number of Civilian Injured']) > 10

In [34]:
#Corner cases (from errors)
cleaned_shooters['Number of Shotguns'].replace('Handgun', 0, inplace=True)
cleaned_shooters['Number of Rifles'].replace('0 (1)', 0, inplace=True)
cleaned_shooters['Number of Handguns'].replace('2 (1)', 0, inplace=True)

In [35]:
#Set gun types to bools
cleaned_shooters['Number of Shotguns'] = pd.to_numeric(cleaned_shooters['Number of Shotguns']) > 0
cleaned_shooters['Number of Rifles'] = pd.to_numeric(cleaned_shooters['Number of Rifles']) > 0
cleaned_shooters['Number of Handguns'] = pd.to_numeric(cleaned_shooters['Number of Handguns']) > 0

In [36]:
#Clean fate of shooter results
cleaned_shooters['Fate of Shooter'].replace('Custody / Escaped', 'Custody/Escaped', inplace=True)

In [37]:
len(cleaned_shooters.loc[
    (cleaned_shooters['Fate of Shooter'].isnull()) | 
    (cleaned_shooters['Fate of Shooter'] == 'FALSE')
])

3

In [38]:
#Drop the three where we don't know the fate of the shooter
cleaned_shooters = cleaned_shooters.loc[
    (~cleaned_shooters['Fate of Shooter'].isnull()) & 
    ~(cleaned_shooters['Fate of Shooter'] == 'FALSE')
]

In [39]:
#Change 'Unknowns' to 'No' for history of mental illness
cleaned_shooters['History of Mental Illness - General'].replace('Unknown', 'No', inplace=True)

#Change to True/False scheme
cleaned_shooters['History of Mental Illness - General'].replace('Yes', True, inplace=True)
cleaned_shooters['History of Mental Illness - General'].replace('No', False, inplace=True)

In [40]:
#We'll iterate over our desired features and flatten them s.t.
#each unique value has a column, and if the incident has that value
#we'll just assign the matching column a 1 and the other ones a 0
marketbasket = cleaned_shooters[[
    'Number of Civilian Fatalities',
    'Number of Civilian Injured',
    'Number of Enforcement Fatalities',
    'Number of Enforcement Injured',
    'Military Experience',
    'Number of Shotguns',
    'Number of Rifles',
    'Number of Handguns',
    'High Civilian Casualties',
    'History of Mental Illness - General'
]]


#Get the logical inverses as well
marketbasket['No Civilian Fatalities'] = ~marketbasket['Number of Civilian Fatalities']
marketbasket['No Civilian Injured'] = ~marketbasket['Number of Civilian Injured']
marketbasket['No Enforcement Fatalities'] = ~marketbasket['Number of Enforcement Fatalities']
marketbasket['No Enforcement Injured'] = ~marketbasket['Number of Enforcement Injured']
marketbasket['No History of Mental Illness - General'] = ~marketbasket['History of Mental Illness - General']


##TODO: Need to finish generating the rest of the dataframe by expanding out unique values across columns
desired_features = [
    'Fate of Shooter',
    'Possible Motive - General'
]

for feature in desired_features:
    for val in cleaned_shooters[feature].unique():
        marketbasket[feature + ' ' + str(val)] = (cleaned_shooters[feature] == val)

marketbasket

C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
C:\Users\spenc\Anaconda3\lib\site-packages\ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexi

Number of Civilian Fatalities  Number of Civilian Injured  \
0                             True                        True   
1                             True                        True   
2                             True                        True   
3                             True                        True   
4                             True                        True   
5                             True                        True   
6                             True                        True   
7                             True                        True   
8                             True                        True   
9                             True                        True   
10                            True                        True   
11                            True                        True   
12                            True                        True   
13                            True                        True   
14                            True                        True   
15                            True                        True   
16                            True                        True   
17                            True                        True   
18                            True                        True   
19                            True                        True   
20                            True                        True   
21                            True                        True   
22                            True                        True   
23                            True                        True   
24                            True                        True   
25                            True                        True   
26                            True                        True   
27                           False                        True   
28                            True                        True   
29                            True                        True   
..                             ...                         ...   
305                           True                        True   
306                          False                        True   
307                           True                       False   
308                           True                        True   
309                          False                        True   
310                           True                       False   
311                           True                       False   
312                           True                        True   
313                           True                        True   
314                           True                        True   
315                           True                        True   
316                           True                        True   
317                           True                        True   
318                           True                        True   
319                           True                        True   
320                          False                        True   
321                           True                       False   
322                           True                       False   
323                           True                       False   
324                           True                        True   
325                           True                        True   
326                           True                        True   
327                           True                        True   
328                          False                        True   
329                           True                        True   
330                          False                        True   
331                           True                        True   
332                          False                        True   
333                           True           

In [70]:
#Apply apriori market basket algorithm
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
apriori_res = apriori(marketbasket, min_support=0.05, use_colnames=True)

In [71]:
#Create display copy of apriori results
apriori_res_display = apriori_res.copy(deep=True)
apriori_res_display['itemsets'] = apriori_res_display['itemsets'].apply(lambda x: ''.join(e + ', ' for e in x))

In [72]:
pd.options.display.max_colwidth = 300
apriori_res_display.sort_values('support')

support  \
734   0.051205   
834   0.051205   
544   0.051205   
829   0.051205   
101   0.051205   
536   0.051205   
981   0.051205   
752   0.051205   
658   0.051205   
546   0.051205   
993   0.051205   
1012  0.051205   
1016  0.051205   
1027  0.051205   
513   0.051205   
665   0.051205   
1037  0.051205   
500   0.051205   
483   0.051205   
530   0.051205   
967   0.051205   
966   0.051205   
555   0.051205   
94    0.051205   
95    0.051205   
866   0.051205   
612   0.051205   
881   0.051205   
599   0.051205   
886   0.051205   
887   0.051205   
893   0.051205   
595   0.051205   
567   0.051205   
564   0.051205   
932   0.051205   
938   0.051205   
949   0.051205   
652   0.051205   
955   0.051205   
556   0.051205   
964   0.051205   
148   0.051205   
1050  0.051205   
1049  0.051205   
472   0.051205   
342   0.051205   
341   0.051205   
338   0.051205   
332   0.051205   
330   0.051205   
1177  0.051205   
1189  0.051205   
318   0.051205   
311   0.051205   
1201  0.051205   
729   0.051205   
1208  0.051205   
288   0.051205   
282   0.051205   
274   0.051205   
759   0.051205   
251   0.051205   
245   0.051205   
244   0.051205   
350   0.051205   
353   0.051205   
715   0.051205   
1131  0.051205   
805   0.051205   
465   0.051205   
671   0.051205   
1077  0.051205   
673   0.051205   
674   0.051205   
409   0.051205   
709   0.051205   
681   0.051205   
855   0.051205   
1115  0.051205   
398   0.051205   
700   0.051205   
200   0.051205   
388   0.051205   
778   0.051205   
384   0.051205   
1114  0.051205   
727   0.054217   
660   0.054217   
677   0.054217   
633   0.054217   
666   0.054217   
669   0.054217   
635   0.054217   
1237  0.054217   
1236  0.054217   
273   0.054217   
302   0.054217   
328   0.054217   
343   0.054217   
204   0.054217   
385   0.054217   
393   0.054217   
407   0.054217   
418   0.054217   
435   0.054217   
437   0.054217   
622   0.054217   
456   0.054217   
476   0.054217   
180   0.054217   
506   0.054217   
522   0.054217   
174   0.054217   
540   0.054217   
542   0.054217   
167   0.054217   
557   0.054217   
584   0.054217   
610   0.054217   
467   0.054217   
535   0.054217   
776   0.054217   
1029  0.054217   
903   0.054217   
925   0.054217   
1103  0.054217   
954   0.054217   
963   0.054217   
965   0.054217   
1154  0.054217   
1151  0.054217   
1145  0.054217   
1143  0.054217   
971   0.054217   
973   0.054217   
974   0.054217   
1138  0.054217   
984   0.054217   
59    0.054217   
1021  0.054217   
768   0.054217   
1032  0.054217   
1039  0.054217   
1047  0.054217   
1124  0.054217   
1059  0.054217   
1072  0.054217   
1122  0.054217   
1116  0.054217   
1089  0.054217   
882   0.054217   
1170  0.054217   
930   0.054217   
868   0.054217   
793   0.054217   
794   0.054217   
795   0.054217   
798   0.054217   
1173  0.054217   
779   0.054217   
1214  0.054217   
813   0.054217   
1216  0.054217   
824   0.054217   
1205  0.054217   
1226  0.054217   
840   0.054217   
845   0.054217   
857   0.054217   
91    0.054217   
1233  0.054217   
860   0.054217   
1225  0.054217   
392   0.057229   
247   0.057229   
1123  0.057229   
255   0.057229   
405   0.057229   
1127  0.057229   
1128  0.057229   
280   0.057229   
1133  0.057229   
375   0.057229   
1215  0.057229   
1140  0.057229   
1197  0.057229   
1196  0.057229   
324   0.057229   
326   0.057229   
1167  0.057229   
331   0.057229   
1222  0.057229   
1113  0.057229   
775   0.057229   
414   0.057229   
643   0.057229   
649   0.057229   
651   0.057229   
836   0.057229   
814   0.057229   
811   0.057229   
667   0.057229   
808   0.057229   
786   0.057229   
691   0.057229   
704   0.057229   
777   0.057229   
243   0.057229   
717   0.057229   
725   0.057229   
873   0.057229   
605   0.057229   
922   0.057229   
560   0.057229   
1079  0.057229   
446   0.057229   
455   0.057229   
1073  0.057229   
1070  0.057229   
1056  0.057229   
479   0.057

In [73]:
#Display full dataset for apriori results, limited to itemsets of <=3 elements
max_rows_orig = pd.options.display.max_rows
pd.options.display.max_rows = 1000
apriori_res.loc[[len(itemset) <= 3 for itemset in apriori_res['itemsets']]]

support  \
0    0.834337   
1    0.807229   
2    0.063253   
3    0.111446   
4    0.138554   
5    0.234940   
6    0.575301   
7    0.153614   
8    0.289157   
9    0.165663   
10   0.192771   
11   0.960843   
12   0.936747   
13   0.710843   
14   0.433735   
15   0.445783   
16   0.102410   
17   0.099398   
18   0.084337   
19   0.292169   
20   0.099398   
21   0.066265   
22   0.081325   
23   0.650602   
24   0.108434   
25   0.129518   
26   0.213855   
27   0.493976   
28   0.150602   
29   0.277108   
30   0.183735   
31   0.804217   
32   0.789157   
33   0.557229   
34   0.400602   
35   0.361446   
36   0.063253   
37   0.093373   
38   0.246988   
39   0.096386   
40   0.063253   
41   0.075301   
42   0.081325   
43   0.120482   
44   0.201807   
45   0.463855   
46   0.153614   
47   0.246988   
48   0.156627   
49   0.780120   
50   0.759036   
51   0.560241   
52   0.310241   
53   0.379518   
54   0.099398   
55   0.084337   
56   0.081325   
57   0.216867   
58   0.084337   
59   0.054217   
60   0.075301   
61   0.099398   
62   0.093373   
63   0.063253   
64   0.075301   
65   0.057229   
66   0.057229   
67   0.069277   
68   0.126506   
69   0.126506   
70   0.069277   
71   0.081325   
72   0.057229   
73   0.117470   
74   0.075301   
75   0.099398   
76   0.210843   
77   0.198795   
78   0.135542   
79   0.117470   
80   0.111446   
81   0.123494   
82   0.201807   
83   0.081325   
84   0.111446   
85   0.548193   
86   0.542169   
87   0.373494   
88   0.289157   
89   0.271084   
90   0.063253   
91   0.054217   
92   0.114458   
93   0.072289   
94   0.051205   
95   0.051205   
96   0.096386   
97   0.141566   
98   0.144578   
99   0.057229   
100  0.102410   
101  0.051205   
102  0.271084   
103  0.268072   
104  0.156627   
105  0.132530   
106  0.093373   
107  0.057229   
108  0.156627   
109  0.147590   
110  0.153614   
111  0.084337   
112  0.180723   
113  0.177711   
114  0.150602   
115  0.123494   
116  0.066265   
117  0.075301   
118  0.915663   
119  0.689759   
120  0.406627   
121  0.433735   
122  0.102410   
123  0.087349   
124  0.081325   
125  0.292169   
126  0.099398   
127  0.063253   
128  0.075301   
129  0.668675   
130  0.391566   
131  0.424699   
132  0.102410   
133  0.090361   
134  0.084337   
135  0.280120   
136  0.093373   
137  0.057229   
138  0.066265   
139  0.277108   
140  0.313253   
141  0.102410   
142  0.078313   
143  0.253012   
144  0.066265   
145  0.117470   
146  0.057229   
147  0.057229   
148  0.051205   
149  0.066265   
150  0.093373   
151  0.075301   
152  0.078313   
153  0.111446   
154  0.186747   
155  0.385542   
156  0.150602   
157  0.237952   
158  0.626506   
159  0.614458   
160  0.412651   
161  0.286145   
162  0.295181   
163  0.060241   
164  0.081325   
165  0.171687   
166  0.081325   
167  0.054217   
168  0.072289   
169  0.096386   
170  0.090361   
171  0.060241   
172  0.072289   
173  0.057229   
174  0.054217   
175  0.066265   
176  0.117470   
177  0.120482   
178  0.063253   
179  0.075301   
180  0.054217   
181  0.111446   
182  0.075301   
183  0.099398   
184  0.195783   
185  0.186747   
186  0.114458   
187  0.108434   
188  0.099398   
189  0.120482   
190  0.189759   
191  0.108434   
192  0.469880   
193  0.466867   
194  0.304217   
195  0.271084   
196  0.213855   
197  0.057229   
198  0.099398   
199  0.072289   
200  0.051205   
201  0.096386   
202  0.138554   
203  0.141566   
204  0.054217   
205  0.102410   
206  0.262048   
207  0.256024   
208  0.144578   
209  0.132530   
210  0.087349   
211  0.057229   
212  0.177711   
213  0.174699   
214  0.144578   
215  0.114458   
216  0.066265   
217  0.075301   
218  0.771084   
219  0.542169   
220  0.379518   
221  0.352410   
222  0.063253   
223  0.084337   
224  0.246988   
225  0.096386   
226  0.060241   
227  0.072289   
228  0.533133   
229  0.367470   
230  0.349398   
231  0.063253   
232  0.084337   
233  0.234940   
234  0.0903

In [63]:
#Reset number of rows that can be shown
pd.options.display.max_rows = max_rows_orig

Great, all of our data is wrapped up in the marketbasket and we have apriori support values for all potential combinations of different variables down to the case where 5% of incidents exhibited a given combination. 

I'm going to start by drilling down into data involving members of the military. It's an interesting twist that someone who would fight to defend the country would turn their gun on the general population. We'll see what shakes out here.

In [74]:
#How many incidents had a shooter w/ military experience?
len(marketbasket.loc[marketbasket['Military Experience'] == True])

37

In [65]:
#Select market basket results with 'Military Experience'
military_apriori = apriori_res.loc[['Military Experience' in elem for elem in apriori_res['itemsets']]]
military_apriori

support  \
3    0.111446   
24   0.108434   
42   0.081325   
60   0.075301   
61   0.099398   
62   0.093373   
63   0.063253   
64   0.075301   
152  0.078313   
168  0.072289   
169  0.096386   
170  0.090361   
171  0.060241   
172  0.072289   
247  0.057229   
248  0.072289   
249  0.069277   
250  0.060241   
322  0.066265   
323  0.060241   
324  0.057229   
325  0.090361   
326  0.057229   
327  0.066265   
328  0.054217   
329  0.060241   
476  0.054217   
477  0.069277   
478  0.066265   
479  0.057229   
538  0.063253   
539  0.057229   
540  0.054217   
541  0.087349   
542  0.054217   
543  0.063253   
544  0.051205   
545  0.057229   
664  0.066265   
665  0.051205   
775  0.057229   
776  0.054217   
777  0.057229   
880  0.063253   
963  0.054217   
964  0.051205   
965  0.054217   

                                                                                                                                            itemsets  
3                                                                                                                              [Military Experience]  
24                                                                                              [Number of Civilian Fatalities, Military Experience]  
42                                                                                                 [Number of Civilian Injured, Military Experience]  
60                                                                                                         [Military Experience, Number of Handguns]  
61                                                                                                  [Military Experience, No Enforcement Fatalities]  
62                                                                                                     [Military Experience, No Enforcement Injured]  
63                                                                                     [Military Experience, No History of Mental Illness - General]  
64                                                                                                   [Military Experience, Fate of Shooter Deceased]  
152                                                                 [Number of Civilian Fatalities, Number of Civilian Injured, Military Experience]  
168                                                                         [Number of Civilian Fatalities, Military Experience, Number of Handguns]  
169                                                                  [Number of Civilian Fatalities, Military Experience, No Enforcement Fatalities]  
170                                                                     [Number of Civilian Fatalities, Military Experience, No Enforcement Injured]  
171                                                     [Number of Civilian Fatalities, Military Experience, No History of Mental Illness - General]  
172                                                                   [Number of Civilian Fatalities, Military Experience, Fate of Shooter Deceased]  
247                                                                            [Number of Civilian Injured, Military Experience, Number of Handguns]  
248                                                                     [Number of Civilian Injured, Military Experience, No Enforcement Fatalities]  
249                                                                        [Number of Civilian Injured, Military Experience, No Enforcement Injured]  
250                                                                      [Number of Civilian Injured, Military Experience, Fate of Shooter Deceased]  
322                                                                             [Military Experience, Number of Handguns, No Enforcement Fatalities]  
323                                                                                [Military Experience, Number of Handguns, No Enforcement Injured]  
324               

In [77]:
#How often do shooters w/o military experience die on the scene?
deceased_no_military_experience = marketbasket.loc[
    (marketbasket['Fate of Shooter Deceased'] == True) &
    (marketbasket['Military Experience'] == False)
]
no_military_experience = marketbasket.loc[marketbasket['Military Experience'] == False]
float(len(deceased_no_military_experience)) / float(len(no_military_experience))

0.4033898305084746

Here's a breakdown of the results from my military experience analysis:
- 37 of the 335 incidents in question involved a shooter with military experience (~11%).
    - According to an analysis by FiveThirtyEight, it's estimated that 7.3% of all living Americans served in the armed forces.
    - I'm going to chalk up that gap in percentage of shooters relative to the overall population to small sample size. There's no way to know exactly how many mass shootings have occurred since 1960, but if we had all of them then I'd guess that this gap would shrink.
- **In 67.5% of incidents where the shooter had military experience, they died on the scene**
    - That's a huge uptick from the general case: 43.4% of all shooters die on the scene
    - 40.3% of shooters with no military experience die on the scene
    - I'll have to run a t-test, but it doesn't seem like a small sample size would explain such a huge uptick.
    - Military experience seems to be a large contributing factor in why someone would die taking a final stand.
- 81% of cases w/ military experience result in officer injury or death
    - General case: 91.6% of cases result in no enforcement injuries or deaths
    - But again, a 10% shift in this subset is a matter of 3-4 cases, probably not enough to go off of

In [64]:
num_officers_not_targeted = len(shooters.loc[(shooters['Number of Enforcement Fatalities'] == 0) & (shooters['Number of Enforcement Injured'] == 0)])
print("Officers were injured or killed ", float(len(shooters) - num_officers_not_targeted) / float (len(shooters)), " of the time.")

Officers were injured or killed  0.08358208955223881  of the time.


Interesting. Nearly every example of a shooting where they had military experience resulted in no officer casualties. That being said, it looks roughly proportional to the number of shooters without military experience that did/didn't target law enforcement. In other words, military experience doesn't appear to make a significant difference in who they choose to target.

In [50]:
from builtins import any as b_any

#Get motive results, remove ones where we don't know their motive or there are multiple ill-defined motives
motive_apriori = apriori_res.loc[[b_any('Possible Motive - General' in e for e in elem) for elem in apriori_res['itemsets']]]
# motive_apriori = motive_apriori[['Possible Motive - General Unknown' not in elem for elem in motive_apriori['itemsets']]]
# motive_apriori = motive_apriori[['Possible Motive - General Multiple motives' not in elem for elem in motive_apriori['itemsets']]]
motive_apriori
# motive_apriori.sort_values('support')

support  \
16   0.292169   
30   0.246988   
42   0.216867   
59   0.114458   
79   0.292169   
84   0.280120   
88   0.253012   
89   0.117470   
100  0.171687   
131  0.246988   
135  0.234940   
138  0.210843   
139  0.108434   
168  0.216867   
172  0.207831   
175  0.189759   
196  0.114458   
200  0.108434   
221  0.280120   
225  0.253012   
226  0.117470   
230  0.246988   
231  0.108434   
254  0.171687   
258  0.162651   
261  0.147590   
295  0.234940   
298  0.210843   
299  0.108434   
302  0.204819   
330  0.207831   
333  0.189759   
336  0.186747   
344  0.108434   
357  0.246988   
358  0.108434   
381  0.162651   
384  0.147590   
387  0.144578   
404  0.204819   
419  0.186747   
432  0.144578   

                                                                                                                                                                                      itemsets  
16                                                                                                                                                         [Possible Motive - General Unknown]  
30                                                                                                                          [Number of Civilian Fatalities, Possible Motive - General Unknown]  
42                                                                                                                             [Number of Civilian Injured, Possible Motive - General Unknown]  
59                                                                                                                                     [Number of Handguns, Possible Motive - General Unknown]  
79                                                                                                                              [No Enforcement Fatalities, Possible Motive - General Unknown]  
84                                                                                                                                 [No Enforcement Injured, Possible Motive - General Unknown]  
88                                                                                                                 [No History of Mental Illness - General, Possible Motive - General Unknown]  
89                                                                                                                               [Fate of Shooter Deceased, Possible Motive - General Unknown]  
100                                                                                             [Number of Civilian Fatalities, Number of Civilian Injured, Possible Motive - General Unknown]  
131                                                                                              [Number of Civilian Fatalities, No Enforcement Fatalities, Possible Motive - General Unknown]  
135                                                                                                 [Number of Civilian Fatalities, No Enforcement Injured, Possible Motive - General Unknown]  
138                                                                                 [Number of Civilian Fatalities, No History of Mental Illness - General, Possible Motive - General Unknown]  
139                                                                                               [Number of Civilian Fatalities, Fate of Shooter Deceased, Possible Motive - General Unknown]  
168                                                                                                 [Number of Civilian Injured, No Enforcement Fatalities, Possible Motive - General Unknown]  
172                                                                                                    [Number of Civilian Injured, No Enforcement Injured, Possible Motive - General Unknown]  
175                                                                                    [Number of Civilian Injured, No History of Mental Illness - General, Possible Motive - General Unkn

The market basket analysis gives us the following insights:
- Law enforcement is injured in just ~6.3% of mass shootings; officers die in 3.9% of incidents.
- Officers were neither injured nor killed in 307 out of 335 incidents (91.6%).
    - Is that discrepancy because law enforcement is adequately trained for these kinds of situations?
    - Or is it because the shooters aren't actively targeting law enforcement?
    - I'm guessing the latter, because the analysis points out that a civilian dies in 83.6% of shootings; if they were targeting law enforcement too, they'd probably have a higher 'success' rate
- 11% of the shooters had military experience
- There's a papertrail for mental illness in only 28.9% of cases

In [48]:
#Group the dataset by year
shooters['Date'] = pd.to_datetime(shooters['Date'])
dategroups = shooters.groupby(shooters.Date.dt.year)


#Iterate over groups, count each gun type
for name, grouped in dategroups:
    gun_counts = grouped['Type of Gun - General'].count().transpose()
    print(name)
    print(gun_counts)
    


1966
2
1971
1
1972
1
1974
2
1976
2
1979
2
1982
2
1983
2
1984
3
1985
2
1986
3
1987
1
1988
6
1989
3
1990
1
1991
5
1992
4
1993
9
1994
4
1995
4
1996
3
1997
6
1998
5
1999
7
2000
1
2001
2
2002
2
2003
3
2004
2
2005
3
2006
6
2007
10
2008
6
2009
6
2010
2
2011
6
2012
14
2013
16
2014
13
2015
65
2016
98


In [49]:
from altair import *
chart = Chart(data).mark_area(
    stacked='center',
).encode(
    color=Color('series:N',
        scale=Scale(
            range='category20b',
        ),
    ),
    x=X('date:T',
        axis=Axis(
            axisWidth=0.0,
            format='%Y',
            labelAngle=0.0,
            tickSize=0.0,
        ),
        scale=Scale(
            nice='month',
        ),
        timeUnit='yearmonth',
    ),
    y=Y('sum(count):Q',
        axis=False,
    ),
).configure_cell(
    height=200.0,
    width=300.0,
)

SchemaValidationError: Invalid specification

        altair.vegalite.v2.schema.core.MarkDef, validating 'additionalProperties'

        Additional properties are not allowed ('stacked' was unexpected)
        

Alright, here's an idea: let's look at the density of different types of shootings across the country based on motive first, then other variables.

In [ ]:
#Let's start by looking at how many examples there are for each motive
shooters['Possible Motive - General'] = shooters['Possible Motive - General'].str.lower()
motive_groups = shooters.groupby('Possible Motive - General')
for name, group in motive_groups:
    print(name, ": ", len(group))

I suppose it's a blessing that there are too few examples of any one kind of shooting to use. In any event, it means that standard density estimation techniques are out the window, except when considering all examples.

In [ ]:
from mpl_toolkits.basemap import Basemap

COLORS=['#1DACE8','#1C366B','#F24D29','#E5C4A1','#C4CFD0']
SELECTED_GROUPS = [
    'social dispute',
    'terminated/denied/reprimanded',
    'domestic dispute',
    'mental illness',
    'political/religious ideals'
]

fig, ax = plt.subplots()
shooter_map = Basemap(llcrnrlon=-145.5,llcrnrlat=1.,urcrnrlon=-2.566,urcrnrlat=46.352,\
            rsphere=(6378137.00,6356752.3142),
            resolution='l',area_thresh=1000.,projection='merc',
            lat_1=50.,lon_0=-107.,ax=ax)
shooter_map.drawcoastlines()
shooter_map.drawcountries()
shooter_map.fillcontinents(color = 'black')
shooter_map.drawmapboundary()


##TODO: fix the coordinates and map projection

#Plot the shooting motivation groups we've indicated above
color_counter = 0
for name, group in motive_groups:
    if name in SELECTED_GROUPS:
        shooter_map.plot(
            group['Latitude'], 
            group['Longitude'],
#             color=COLORS[color_counter],
            'bo',
            markersize=20
        )
        color_counter += 1

plt.show()